Migrate BraviaTV to new async backend

Drafteed opened the pull request:
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This PR will swap the [Sony Bravia TV](https://www.home-assistant.io/integrations/braviatv/) integration backend from unmaintained [bravia-tv](https://github.com/dcnielsen90/python-bravia-tv) to new [pybravia](https://github.com/Drafteed/pybravia). The new library has been completely rewritten using asynchronous calls and aiohttp. This is my first experience of creating a python library and if there are any comments regarding it, I have created a [separate issue in the library repository](https://github.com/Drafteed/pybravia/issues/1).

**Speed comparison (integration startup time/Apple M1):**

Old backend: **3.6s**
New backend: **1.6s** (**1.09s** if PSK auth used)

**Todo list for integration in the future:**
- [ ] Add new autorization method with PSK (Pre-Shared-Key) in `config_flow` (already supported by backend)
- [ ] Add `Button` platform for reboot TV and terminate apps
- [ ] Add `play_media` support
- [ ] Add support `media_browser` and separate channels from `source_list` ([related issue](https://github.com/home-assistant/core/issues/70886))
- [ ] Add `ssdp` discovery.

**Latest changelog of pybravia:**
https://github.com/Drafteed/pybravia/commits/v0.2.0

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [x] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #72345, fixes #74658
- This PR is related to issue: #70886
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

Drafteed committed:
new backend lib

Drafteed committed:
change source_map to dict, tweaks

Drafteed committed:
bump pybravia to 0.1.2

Drafteed committed:
fix manifest logger name

MartinHjelmare commented on homeassistant/components/braviatv/manifest.json:
```diff
@@ -2,9 +2,9 @@
   "name": "Sony Bravia TV",
   "documentation": "https://www.home-assistant.io/integrations/braviatv",
-  "requirements": ["bravia-tv==1.0.11"],
+  "requirements": ["pybravia==0.1.2"],
```
The library traps exceptions now and logs warnings.

https://github.com/Drafteed/pybravia/blob/6dc30cfc41fed5a7b1d1bc6fb8a8bd90737a32d7/pybravia/braviatv.py#L126

It's better to raise library specific exceptions on error instead and let the library user handle error/warning logging. It's easy to spam the log otherwise and it's easier for the library user to be able to take action on error if the library raises an exception. Debug logs are of course always ok.

MartinHjelmare replied:
I'd recommend to add type annotations to the library and declare it to be typed with a `py.typed` file.

MartinHjelmare commented on homeassistant/components/braviatv/config_flow.py:
```diff
@@ -138,26 +137,19 @@ class BraviaTVOptionsFlowHandler(config_entries.OptionsFlow):
-            braviarc.load_source_list
-        )
-        self.source_list = {item: item for item in content_mapping}
+        assert coordinator is not None
```
Add a type annotation above instead to the coordinator variable.

MartinHjelmare commented on homeassistant/components/braviatv/config_flow.py:
```diff
@@ -138,26 +137,19 @@ class BraviaTVOptionsFlowHandler(config_entries.OptionsFlow):
         self, user_input: dict[str, Any] | None = None
     ) -> FlowResult:
         """Manage the options."""
         coordinator = self.hass.data[DOMAIN][self.config_entry.entry_id]
```
```suggestion
        coordinator: BraviaTVCoordinator = self.hass.data[DOMAIN][self.config_entry.entry_id]
```

Drafteed committed:
tweaks from review

bieniu commented:
I started messing around with this change and I'm having trouble connecting using PIN, investigating.
```
2022-08-02 10:40:54.992 DEBUG (MainThread) [pybravia.braviatv] Connect with pin: 4823, psk: False, clientid: HomeAssistant, nickname: Home Assistant
2022-08-02 10:40:54.992 DEBUG (MainThread) [pybravia.braviatv] Request http://192.168.2.45/sony/accessControl, data: {'method': 'actRegister', 'params': [{'clientid': 'HomeAssistant', 'nickname': 'Home Assistant', 'level': 'private'}, [{'value': 'yes', 'function': 'WOL'}]], 'id': 1, 'version': '1.0'}, headers: {'Authorization': 'Basic OjAwMDA=', 'Connection': 'keep-alive'}
2022-08-02 10:40:55.553 DEBUG (MainThread) [pybravia.braviatv] Response status: 200, result: {'result': [], 'id': 1}
2022-08-02 10:40:55.553 DEBUG (MainThread) [pybravia.braviatv] Request http://192.168.2.45/sony/system, data: {'method': 'setWolMode', 'params': [{'enabled': True}], 'id': 1, 'version': '1.0'}, headers: {}
2022-08-02 10:40:55.563 DEBUG (MainThread) [pybravia.braviatv] Response status: 403, result: {}
2022-08-02 10:40:55.563 DEBUG (MainThread) [pybravia.braviatv] Connect status: False
```

bieniu commented on homeassistant/components/braviatv/__init__.py:
```diff
@@ -110,178 +112,165 @@ def __init__(
+            self.media_content_id = None
+            self.media_content_type = None
+        if not playing_info:
+            self.media_title = "App"
```
I wonder if we should add here:
```python
self.media_content_type = MEDIA_TYPE_APP
```

bieniu replied:
And maybe we should change `App` to `SmartTV`? We really don't know what is running when `playing_info is None`.

Drafteed replied:
> And maybe we should change `App` to `SmartTV`? We really don't know what is running when `playing_info is None`.

It won't be a breaking change? :)

Drafteed committed:
tweaks from review and discussion

MartinHjelmare commented on homeassistant/components/braviatv/manifest.json:
```diff
@@ -2,7 +2,7 @@
   "name": "Sony Bravia TV",
   "documentation": "https://www.home-assistant.io/integrations/braviatv",
-  "requirements": ["pybravia==0.1.3"],
+  "requirements": ["pybravia==0.1.5"],
```
It looks like the library is defaulting to not raising exceptions.

Drafteed replied:
By default, no. I thought that the result of executing the connect method (returns bool) is enough to understand whether it was connected or not. If necessary, I can throw an exception by default, but do I need to handle it somehow on the HA side?

Drafteed replied:
@MartinHjelmare Now the library raising exceptions on connect by default. Following [this guide](https://developers.home-assistant.io/docs/integration_fetching_data/#coordinated-single-api-poll-for-data-for-all-entities) I raise `UpdateFailed` in `_async_update_data` on error. Please take a look.

Drafteed committed:
added exceptions handling

Drafteed committed:
refresh after select source

MartinHjelmare commented on homeassistant/components/braviatv/__init__.py:
```diff
@@ -110,178 +115,171 @@ def __init__(
-        power_status = self.braviarc.get_power_status()
+    async def _async_update_data(self) -> None:
+        """Connect and fetch data."""
+        power_status = await self.client.get_power_status()
```
All client request calls should raise an exception if the request fails. So we should handle possible exceptions in all these client calls.

To not have to repeat the code that handles the exception in all service handlers etc we often use a decorator.

Example:
https://github.com/home-assistant/core/blob/b7b965c9c9575e3717eb4bef39809c55c700b7d4/homeassistant/components/vlc_telnet/media_player.py#L50-L68

https://github.com/home-assistant/core/blob/b7b965c9c9575e3717eb4bef39809c55c700b7d4/homeassistant/components/vlc_telnet/media_player.py#L242-L245

Drafteed committed:
tweaks from review, refactor

Drafteed commented:
Martin, thanks for the tips! I changed the structure of the library - it now throws exceptions on all client requests. As you said, I wrapped all the client methods of the coordinator with a decorator + refactored some part of the code on both sides.
In addition, I added basic type annotations to the library.

I tested everything for a couple of days - everything works properly with my XH90.

Note: CI failure is unrelated to these changes.

Drafteed committed:
oops

Drafteed committed:
trigger CI

Drafteed committed:
add coordinator.py to .coveragerc

MartinHjelmare commented on homeassistant/components/braviatv/coordinator.py:
```diff
@@ -0,0 +1,270 @@
+        try:
+            await func(self, *args, **kwargs)
+        except BraviaTVError as err:
+            _LOGGER.error("Command error: %s", err)
```
Should we set the coordinator as not connected here and request a refresh?

Drafteed replied:
I wouldn't change "connected" here, as a conditionally unsupported command sent via the `remote.send_command` would result in a state change. 

In addition, after each command, a refresh request is called and in case of a fail, it will change the connection state:

https://github.com/home-assistant/core/blob/1c2aedc17d6f3a36f1630443aad112c86c4a25b9/homeassistant/components/braviatv/coordinator.py#L119-L126

https://github.com/home-assistant/core/blob/1c2aedc17d6f3a36f1630443aad112c86c4a25b9/homeassistant/components/braviatv/coordinator.py#L196-L200

But I think moving the call `await self.async_request_refresh()` to a decorator to reduce the number of code repeats a good idea.

Drafteed replied:
Done

MartinHjelmare commented on homeassistant/components/braviatv/config_flow.py:
```diff
@@ -138,26 +130,20 @@ class BraviaTVOptionsFlowHandler(config_entries.OptionsFlow):
+
+        await coordinator.async_update_sources()
+        sources = coordinator.source_map.values()
+        self.source_list = [item.get("title", "") for item in sources]
```
What does an empty string as source mean?

Drafteed committed:
tweaks from review

Drafteed commented on homeassistant/components/braviatv/config_flow.py:
```diff
@@ -44,27 +46,16 @@ class BraviaTVConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
         """Initialize."""
-        self.braviarc: BraviaRC | None = None
-        self.host: str | None = None
+        self.client: BraviaTV
```
Done + slightly refactored the storage of config data.

MartinHjelmare commented on homeassistant/components/braviatv/config_flow.py:
```diff
@@ -102,15 +108,12 @@ async def async_step_authorize(
         if user_input is not None:
             try:
-                await self.init_device(user_input[CONF_PIN])
+                self.device_config[CONF_PIN] = user_input[CONF_PIN]
```
Please only wrap the line that can raise in the try... except block. We can move this line up.

MartinHjelmare commented on homeassistant/components/braviatv/coordinator.py:
```diff
@@ -41,6 +41,7 @@ async def wrapper(
         """Catch BraviaTV errors and log message."""
         try:
             await func(self, *args, **kwargs)
+            await self.async_request_refresh()
```
I'd move this down to the end of the body, as we don't expect it to raise.

Drafteed committed:
tweaks from review

MartinHjelmare reviewed (approved):
Nice!

bieniu reviewed (approved):
Tested with my TV and everything works as it should 👍

Drafteed commented:
Thanks, @bieniu 👍
Thanks, @MartinHjelmare 👍

Drafteed commented:
@MartinHjelmare, what do you think about move host validation helper into [/homeassistant/util/network.py:](https://github.com/home-assistant/core/blob/dev/homeassistant/util/network.py)
https://github.com/home-assistant/core/blob/19295d33ba60bad0a9dfbc4d512e156a09a010fc/homeassistant/components/braviatv/config_flow.py#L33-L39
https://github.com/home-assistant/core/blob/19295d33ba60bad0a9dfbc4d512e156a09a010fc/homeassistant/components/brother/config_flow.py#L27-L35
https://github.com/home-assistant/core/blob/19295d33ba60bad0a9dfbc4d512e156a09a010fc/homeassistant/components/dunehd/config_flow.py#L18-L28
https://github.com/home-assistant/core/blob/19295d33ba60bad0a9dfbc4d512e156a09a010fc/homeassistant/components/vilfo/config_flow.py#L32-L39

MartinHjelmare commented:
Sounds good. When would the regex check apply?

Drafteed commented:
> Sounds good. When would the regex check apply?

I think it should be like this:
```py
def is_host_valid(host: str) -> bool:
    """Check if a given string is an IP address or valid hostname."""
    if is_ip_address(host):
        return True
    if host[-1] == ".":
        host = host[:-1]
    allowed = re.compile(r"(?!-)[A-Z\d\-]{1,63}(?<!-)$", re.IGNORECASE)
    return all(allowed.match(x) for x in host.split("."))
```
